### PR TITLE
Performance improvements for SSMS query while trying to expand stored procedures

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

Currently, the SSMS query to expand stored procedures is found to be very slow (~ 6 hours for 1000 procedures). 
The bad performance is found to be majorly occuring from sys.all_objects view. So, the following modifications have been done to the sys.all_objects : 
1. Bad row estimation arising due to the is_ms_shipped column in sys.all_objects view. This change fixes it by pushing down the computation of is_ms_shipped column to the respective unions. 
2. Made the datatype of `type` and `name` column consistent and removed the outer CAST as we are already casting them in the respective UNIONs.
3. Removed subqueries and replaced them with suitable JOINs for better query planning.

Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

### Issues Resolved

BABEL-4008

### Test Scenarios Covered ###
* **Use case based -** Tests for sys.all_objects are already present in `sys_all_objects-vu-*` files

* **Boundary conditions -**  N/A

* **Arbitrary inputs -** N/A

* **Negative test cases -** N/A

* **Minor version upgrade tests -** N/A

* **Major version upgrade tests -** N/A

* **Performance tests -** Evaluated the performance before and after making the changes. Here are the results for 1000 stored procedures.

Before : Execution time = ~ 9 mins , [Query Plan](https://explain.dalibo.com/plan/0405h5490d395148)
After : Execution time = ~ 17 secs , [Query Plan](https://explain.dalibo.com/plan/g1e4e8a84h6gb3d9)

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 1]
```diff
@@ -1268,108 +1268,167 @@ GRANT SELECT ON sys.shipped_objects_not_in_sys TO PUBLIC;
...
+  , 1::sys.bit as is_ms_shipped
+  , 0 as is_published
+  , 0 as is_schema_published
+from pg_class t inner join pg_namespace s on s.oid = t.relnamespace
+left join sys.babelfish_namespace_ext ext on (s.nspname = ext.nspname and ext.dbid = sys.db_id())
+left join sys.shipped_objects_not_in_sys nis on nis.name = t.relname and nis.schemaid = s.oid and nis.type = 'V'
+where t.relkind = 'v'
+and (s.nspname = 'sys' or (nis.name is not null and ext.nspname is not null))
+and has_schema_privilege(s.oid, 'USAGE')
+and has_table_privilege(quote_ident(s.nspname) ||'.'||quote_ident(t.relname), 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER')
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Can we call has_table_privilege with t.oid. The string creation seems redundant

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 2]
```diff
@@ -1268,108 +1268,167 @@ GRANT SELECT ON sys.shipped_objects_not_in_sys TO PUBLIC;
...
-and (s.oid in (select schema_id from sys.schemas) or s.nspname = 'sys')
-and not sys.is_table_type(t.oid)
+and s.nspname <> 'sys' and nis.name is null
+and ext.nspname is not null
+and tt.typrelid is null
 and has_schema_privilege(s.oid, 'USAGE')
 and has_table_privilege(t.oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER')
+ 
+union all
+-- details of user defined and system views
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: Lets add comment on these additional unions on why they are needed

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 3]
```diff
@@ -1411,20 +1470,114 @@ select
...
+left join sys.babelfish_namespace_ext ext on (s.nspname = ext.nspname and ext.dbid = sys.db_id())
+left join sys.shipped_objects_not_in_sys nis on nis.name = p.proname and nis.schemaid = s.oid 
+and nis.type = (case p.prokind
+      when 'p' then 'P'::char(2)
+      when 'a' then 'AF'::char(2)
+      else
+        case 
+          when t.typname = 'trigger'
+            then 'TR'::char(2)
+          when p.proretset then
```
Comment (KushaalShroff) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:1567]: Is this valid for TSQL?

**Review by KushaalShroff — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 3 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:1567]: Yes.

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 1 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: done

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 2 above)
Comment (basasairohan) [reply contrib/babelfishpg_tsql/sql/sys_views.sql]: done

**Review by KushaalShroff — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 4]
```diff
@@ -1268,108 +1268,245 @@ GRANT SELECT ON sys.shipped_objects_not_in_sys TO PUBLIC;
...
+  , 'U'::char(2) as type
+  , 'USER_TABLE' as type_desc
+  , null::timestamp as create_date
+  , null::timestamp as modify_date
+  , 1::sys.bit as is_ms_shipped
+  , 0 as is_published
+  , 0 as is_schema_published
+from pg_class t inner join pg_namespace s on s.oid = t.relnamespace
+left join sys.table_types_internal tt on t.oid = tt.typrelid
+left join sys.babelfish_namespace_ext ext on (s.nspname = ext.nspname and ext.dbid = sys.db_id())
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:1305]: can we directly join pg_class with babelfish_namespace_ext on line 1303?

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**